File: R/04_weekend_weekday.R
# 04_weekend_weekday.R
# Cyclistic Bike-Share Case Study — R Module
#
# Weekend vs weekday ride distribution by rider type.
# Confirms: casual riders 1.61x more likely to ride on weekends than members.
#
# Introduces: lubridate::wday(), if_else(), grouped bar chart

library(tidyverse)

csv_folder <- file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
                        "Data", "DATAfile_Consolidated")
csv_files  <- list.files(csv_folder, pattern = "\\.csv$", full.names = TRUE)
trips      <- read_csv(csv_files, show_col_types = FALSE)

# ---------------------------------------------------------------------------
# Classify each ride as weekend or weekday
#
# wday() is lubridate's day-of-week function.
# Default: 1 = Sunday, 7 = Saturday (same as SQL Server's DATEPART(WEEKDAY))
# DuckDB used DAYOFWEEK() where 0 = Sunday, 6 = Saturday — different numbering,
# same days. Always check the reference day when switching tools.
#
# if_else() is dplyr's vectorised ifelse — cleaner than base R's ifelse()
# for working inside mutate().
# ---------------------------------------------------------------------------

trips <- trips |>
  mutate(
    day_of_week = wday(started_at),                          # 1=Sun, 7=Sat
    is_weekend  = if_else(day_of_week %in% c(1, 7), "Weekend", "Weekday")
  )

# ---------------------------------------------------------------------------
# Summarise: rides by rider type and day type
# ---------------------------------------------------------------------------

day_summary <- trips |>
  group_by(member_casual, is_weekend) |>
  summarise(rides = n(), .groups = "drop") |>
  group_by(member_casual) |>
  mutate(pct = round(rides / sum(rides) * 100, 1))

cat("--- Weekend vs Weekday Breakdown ---\n")
print(day_summary)

# Weekend skew ratio (casual weekend % / member weekend %)
casual_wknd <- day_summary$pct[day_summary$member_casual == "casual"  & day_summary$is_weekend == "Weekend"]
member_wknd <- day_summary$pct[day_summary$member_casual == "member"  & day_summary$is_weekend == "Weekend"]

cat("\nCasual weekend share:", casual_wknd, "%\n")
cat("Member weekend share:", member_wknd, "%\n")
cat("Weekend skew ratio (casual / member):", round(casual_wknd / member_wknd, 2), "\n")

# ---------------------------------------------------------------------------
# Grouped bar chart
#
# position = "dodge" places bars side by side instead of stacked.
# ---------------------------------------------------------------------------

ggplot(day_summary, aes(x = is_weekend, y = pct, fill = member_casual)) +
  geom_col(position = "dodge", width = 0.6) +
  geom_text(
    aes(label = paste0(pct, "%")),
    position = position_dodge(width = 0.6),
    vjust    = -0.4,
    size     = 4
  ) +
  scale_fill_manual(
    values = c("casual" = "#F4A261", "member" = "#2A9D8F"),
    labels = c("Casual", "Member")
  ) +
  scale_y_continuous(limits = c(0, 85), labels = scales::percent_format(scale = 1)) +
  labs(
    title    = "Casual Riders Skew Heavily Toward Weekends",
    subtitle = "Casual riders are 1.61x more likely to ride on weekends than members",
    x        = NULL,
    y        = "Share of Rider Type's Trips",
    fill     = NULL,
    caption  = "Source: Divvy Trip Data"
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "top")

ggsave(
  file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
            "Visuals", "R", "04_weekend_weekday.png"),
  width = 8, height = 6, dpi = 150
)

File: Python/polars/04_weekend_weekday.py
"""
04_weekend_weekday.py
---------------------
Do members and casual riders show different weekend vs weekday patterns?

Expected output (matching SQL Server, DuckDB, R, and pandas):
  Casual weekend/weekday ratio: ~1.61x

Polars notes:
  - .dt.weekday() returns 1=Monday through 7=Sunday (ISO 8601).
    This differs from pandas (.dt.dayofweek: 0=Mon, 6=Sun) and
    DuckDB (DAYOFWEEK: 0=Sun, 6=Sat). Always check the reference day.
  - Weekend in Polars: weekday >= 6 (Saturday=6, Sunday=7).
  - pl.when().then().otherwise() creates conditional columns inline.
"""

import polars as pl
from utils import load_trips

trips = load_trips()

# Polars .dt.weekday(): 1=Mon, 2=Tue, ..., 6=Sat, 7=Sun
trips = trips.with_columns(
    pl.when(pl.col("started_at").dt.weekday() >= 6)
    .then(pl.lit("weekend"))
    .otherwise(pl.lit("weekday"))
    .alias("day_type")
)

wk = (
    trips
    .group_by(["member_casual", "day_type"])
    .agg(pl.len().alias("rides"))
    .sort(["member_casual", "day_type"])
)

print("=" * 55)
print("  Weekend vs Weekday Rides by Rider Type")
print("=" * 55)
print(f"  {'Rider':<10} {'Day Type':<10} {'Rides':>10}")
print(f"  {'-'*35}")
for row in wk.iter_rows(named=True):
    print(f"  {row['member_casual']:<10} {row['day_type']:<10} {row['rides']:>10,}")

# Ratio
def get_rides(rider, day):
    return wk.filter(
        (pl.col("member_casual") == rider) & (pl.col("day_type") == day)
    )["rides"][0]

for rider in ["member", "casual"]:
    wkend = get_rides(rider, "weekend")
    wkday = get_rides(rider, "weekday")
    ratio = wkend / wkday
    print(f"\n  {rider.capitalize()} weekend/weekday ratio: {ratio:.2f}x")

casual_ratio = get_rides("casual", "weekend") / get_rides("casual", "weekday")
print(f"\n  Casual 1.61x check: {'✓' if abs(casual_ratio - 1.61) < 0.02 else '✗'}")
print("=" * 55)
